iree/hal/local/elf/elf_module_run.c

```diff
@@ -0,0 +1,145 @@
+
+  void* query_fn_ptr = NULL;
+  IREE_RETURN_IF_ERROR(iree_elf_module_lookup_export(
+      &module, IREE_HAL_EXECUTABLE_LIBRARY_EXPORT_NAME, &query_fn_ptr));
+
+  union {
+    const iree_hal_executable_library_header_t** header;
+    const iree_hal_executable_library_v0_t* v0;
+  } library;
+  library.header =
+      (const iree_hal_executable_library_header_t**)iree_elf_call_p_ip(
+          query_fn_ptr, IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION,
+          /*reserved=*/NULL);
+  if (library.header == NULL) {
+    return iree_make_status(IREE_STATUS_NOT_FOUND, "library header is empty");
+  }
+
+  const iree_hal_executable_library_header_t* header = *library.header;
+  if (header->version != IREE_HAL_EXECUTABLE_LIBRARY_VERSION_0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "library version error");
+  }
+
+  if (strncmp(header->name, "simple_mul_dispatch_0", strlen(header->name)) !=
+      0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "library name mismatches");
+  }
+
+  if (library.v0->entry_point_count != 1) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "entry point count mismatches");
+  }
+
+  float arg0[4] = {1.0f, 2.0f, 3.0f, 4.0f};
+  float arg1[4] = {100.0f, 200.0f, 300.0f, 400.0f};
+  float ret0[4] = {0.0f, 0.0f, 0.0f, 0.0f};
+  size_t binding_lengths[3] = {
+      sizeof(arg0),
+      sizeof(arg1),
+      sizeof(ret0),
+  };
+  void* binding_ptrs[3] = {
+      arg0,
+      arg1,
+      ret0,
+  };
+  iree_hal_vec3_t workgroup_count = {{1, 1, 1}};
+  iree_hal_vec3_t workgroup_size = {{1, 1, 1}};
+  iree_hal_executable_dispatch_state_v0_t dispatch_state;
+  memset(&dispatch_state, 0, sizeof(dispatch_state));
+  dispatch_state.workgroup_count = workgroup_count;
+  dispatch_state.workgroup_size = workgroup_size;
+  dispatch_state.binding_count = 1;
+  dispatch_state.binding_lengths = binding_lengths;
+  dispatch_state.binding_ptrs = binding_ptrs;
+  iree_hal_vec3_t workgroup_id = {{0, 0, 0}};
+  int ret = iree_elf_call_i_pp((const void*)library.v0->entry_points[0],
+                               (void*)&dispatch_state, (void*)&workgroup_id);
+
+  if (ret != 0) {
+    return iree_make_status(IREE_STATUS_INTERNAL, "elf call fails");
+  }
+  float expected_ret[4] = {100.0f, 400.0f, 900.0f, 1600.0f};
+  for (int i = 0; i < 4; ++i) {
+    if (ret0[i] != expected_ret[i]) {
+      fprintf(stderr, "ret[%d] is incorrect; expected: %.1f; actual: %1.f\n", i,
+              expected_ret[i], ret0[i]);
+      ret_status = iree_make_status(IREE_STATUS_INTERNAL, "output mismatches");
+    }
+  }
+
+  iree_elf_module_deinitialize(&module);
+  return ret_status;
+}
+
+int main() {
+  const iree_status_t result = Run();
+  if (!iree_status_is_ok(result)) {
+    char* message;
+    size_t message_length;
+    iree_status_to_string(result, &message, &message_length);
```

Comment:
Ahh... yes, I was thinking about not using `iree::base::status` (because of the abseil dependency) but `iree_status_fprintf` is in `iree::base`. Fixed.